lib/common/messenger.js

```diff
@@ -263,77 +264,103 @@ function messengerFactory(
+	    context.reject = reject;
             self.receive.queue(queueName(name, routingKey), queueOptions, function (q) {
-                q.subscribe(subscribeOptions, function (data, headers, deliveryInfo) {
-                    clearTimeout(timeout);
-
-                    if (subscription) {
-                        subscription.dispose();
-                    }
-
-                    statsd.timing(
-                        statsd.sanitize(
-                            'messenger.%s.%s'.format(
-                                name, routingKey
-                            )
-                        ),
-                        timer.stop()
-                    );
-
-                    var message = new Message(data, headers, deliveryInfo);
-
-                    if (message.data instanceof ErrorEvent) {
-                        reject(message.data);
-                    } else {
-                        if (_.isFunction(type) && !(message.data instanceof type)) {
-                            reject(
-                                new Error(
-                                    'Invalid Response Type (%s:%s:%s).'.format(
-                                        name,
-                                        routingKey,
-                                        type || 'Object'
-                                    )
-                                )
-                            );
-                        } else {
-                            resolve(message.data);
-                        }
-                    }
-                }).addCallback(function (options) {
-                    timeout = setTimeout(function () {
-                        clearTimeout(timeout);
-
-                        if (subscription) {
-                            subscription.dispose();
-                        }
-
-                        reject(
-                            new Errors.RequestTimedOutError(
-                                'Request Timed Out (%s:%s:%s).'.format(
-                                    name,
-                                    routingKey,
-                                    type || 'Object'
-                                )
-                            )
-                        );
-                    }, self.timeout);
-
-                    subscription = new Subscription(q, options);
-
-                    timer.start();
-
-                    self.publish(name, routingKey, data, {
-                        replyTo: q.name
-                    });
-                });
+		context.queue = q;
+                q.subscribe(
+		    subscribeOptions, 
+		    self.subscribeTimeout.bind(self, context)
+		)              
+		.addCallback(self.subscribeCallback.bind(self, context));
             });
         });
     };
 
+    Messenger.prototype.subscribeTimeout = function(context, data, headers, deliveryInfo) {
+	clearTimeout(context.timeout);
+
+	if (context.subscription) {
+	    if (context.subscription._disposed === false) {
```

Comment:
You could combine these two if statements into one, since this code appears in more than one location I recommend you make it into its own method.
```
safeSubscriptionDispose: function (context) {
  context.subscription && !context.subscription._disposed && context.subscription.dispose();
}
```